File: dsa_algo/patterns/fast_slow_pointers/6_rearrange_list.cpp
#include<bits/stdc++.h>
class list_node{
    public:
        int value = 0;
        list_node *next;
        list_node(int value)
        {
            this->value = value;
            next = nullptr;
        }
};

class rearrange_list {
    public:
        static void reorder(list_node *head)
        {
            if(head == nullptr || head->next == nullptr)
                return;
            list_node *slow = head;
            list_node *fast = head;
            while(fast != nullptr && fast->next != nullptr)
            {
                slow = slow->next;
                fast = fast->next->next;
            }
            list_node *head_second_half = reverse(slow);
            list_node *head_first_half = head;
            while(head_first_half != nullptr && head_second_half != nullptr)
            {
                list_node *temp = head_first_half->next;
                head_first_half->next = head_second_half;
                head_first_half = temp;
                temp = head_second_half->next;
                head_second_half->next = head_first_half;
                head_second_half = temp;
            }
            if(head_first_half != nullptr)
                head_first_half->next = nullptr;
        }
    private:
        static list_node *reverse(list_node *head)
        {
            list_node *prev = nullptr;
            while(head != nullptr)
            {
                list_node *next = head->next;
                head->next = prev;
                prev = head;
                head = next;
            }
            return prev;
        }
};

int main(int argc, char *argv[])
{
    list_node *head = new list_node(2);
    head->next = new list_node(4);
    head->next->next = new list_node(6);
    head->next->next->next = new list_node(8);
    head->next->next->next->next = new list_node(10);
    head->next->next->next->next->next = new list_node(12);
    rearrange_list::reorder(head);
    while(head != nullptr)
    {
        std::cout<<head->value<<" ";
        head = head->next;
    }
}

File: dsa_algo/patterns/fast_slow_pointers/6_rearrange_list.py
from __future__ import print_function

class node:
    def __init__(self, value, next = None):
        self.value = value
        self.next = next
    def print_list(self):
        temp = self
        while temp is not None:
            print(str(temp.value) + " ", end='')
            temp = temp.next
        print()
def reorder(head):
    if head is None or head.next is None:
        return
    slow, fast = head, head
    while fast is not None and fast.next is not None:
        slow = slow.next
        fast = fast.next.next
    head_second_half = reverse(slow)
    head_first_half = head
    while head_first_half is not None and head_second_half is not None:
        temp = head_first_half.next
        head_first_half.next = head_second_half
        head_first_half = temp

        temp =  head_second_half.next
        head_second_half.next = head_first_half
        head_second_half = temp
    if head_first_half is not None:
        head_first_half.next = None
def reverse(head):
    prev = None
    while head is not None:
        next = head.next
        head.next = prev
        prev = head
        head = next
    return prev

def main():
    head = node(2)
    head.next = node(4)
    head.next.next = node(6)
    head.next.next.next = node(8)
    head.next.next.next.next = node(10)
    head.next.next.next.next.next = node(12)
    reorder(head)
    head.print_list()

main()
